File: 2020/Day 03/1.py
Map, Slopes = {(x, y): int(Ch == '#') for (y, Ln) in
    enumerate(open('input.txt')) for (x, Ch) in enumerate(Ln)
if Ch != '\n'}, [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
Mx, My = max(Map.keys()); from math import prod

def Solve(A, B):
    S = X = Y = 0
    while Y <= My:
        S += Map[X, Y]
        X, Y = (X + A) % \
            (Mx + 1), Y + B
    return S

print("Silver:", Solve(*Slopes[1]))
print("Gold:", prod([Solve(*v) for v in Slopes]))

File: 2020/Day 03/1.hs
{-# LANGUAGE TupleSections #-}
import Text.Printf
import Data.Maybe
import Data.List

main :: IO ()
main = do
  let slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
  trees <- sequence (solve <$> slopes) <$> input
  printf "Silver: %d\n" $trees !!1
  printf "Gold: %d\n" $product trees

solve :: (Int, Int) -> [[Int]] -> Int
solve (x, y) ln = sum $uncurry (!!)
  <$> zip (every y ln) [0, x..]

every :: Int -> [a] -> [a]
every n = unfoldr$ \xs-> fmap
  (, drop n xs) (listToMaybe xs)

input :: IO [[Int]]
input = map (cycle.map (fromEnum.(== '#')))
  .lines <$> readFile "input.txt"
